### PR TITLE
Rework the test hierarchy and make the TCK more consistent with HV testing

### PR DESCRIPTION
**Pull request by gsmet**
Note that it also includes a fix for BVTCK-125.

Once integrated, the HV CI will fail until all related fixes in HV are in. Don't worry about it, it's all under control!

https://hibernate.atlassian.net/browse/BVTCK-121

### CODE REVIEW

**Comment by gunnarmorling** [comment]
> Don't worry about it, it's all under control!

:)

File: tests/src/main/java/org/hibernate/beanvalidation/tck/util/ConstraintViolationAssert.java [hunk 1]
```diff
@@ -45,29 +45,37 @@
...
 			}
 
-			assertContainsOnly( actualPaths, paths );
+			Assertions.assertThat( actualPaths ).containsOnly( paths );
 		}
 
-		public static void assertContainsPath(Set<? extends ConstraintViolation<?>> actualViolations, PathExpectation expectedPath) {
-			assertNotNull( expectedPath );
+		public void containsPath(PathExpectation expectedPath) {
+			isNotNull();
```
Comment (gunnarmorling) [inline tests/src/main/java/org/hibernate/beanvalidation/tck/util/ConstraintViolationAssert.java]: Is the parameter missing?

File: tests/src/main/java/org/hibernate/beanvalidation/tck/tests/constraints/builtinconstraints/FuturePastConstraintsTest.java [hunk 2]
```diff
@@ -422,6 +482,28 @@ public PastOrPresentDummyEntity(ZonedDateTime dateTime) {
 		}
 	}
 
+	public class PastRelativePartialDummyEntity {
```
Comment (gunnarmorling) [inline tests/src/main/java/org/hibernate/beanvalidation/tck/tests/constraints/builtinconstraints/FuturePastConstraintsTest.java]: Where are they coming from? Seems a bit unrelated to the rest of this commit?

File: tests/src/main/java/org/hibernate/beanvalidation/tck/tests/constraints/builtinconstraints/DigitsConstraintTest.java [hunk 3]
```diff
@@ -55,7 +55,7 @@ public void testDigitsConstraint() {
 
 		Set<ConstraintViolation<DigitsDummyEntity>> constraintViolations = validator.validate( dummy );
 		// only the max constraints on the primitive values should fail. Object values re still null and should pass per spec
-		assertCorrectNumberOfViolations( constraintViolations, 4 );
+		assertNumberOfViolations( constraintViolations, 4 );
 		assertCorrectPropertyPaths(
```
Comment (gunnarmorling) [inline tests/src/main/java/org/hibernate/beanvalidation/tck/tests/constraints/builtinconstraints/DigitsConstraintTest.java]: This remains as a short code for actual properties?

File: tests/src/main/java/org/hibernate/beanvalidation/tck/tests/methodvalidation/ValidateParametersTest.java [hunk 4]
```diff
@@ -236,22 +226,17 @@ public void testCrossParameterConstraintGivenSeveralTimes() throws Exception {
...
-		assertCorrectPathNodeNames(
-				violations,
-				names( methodName, TestUtil.CROSS_PARAMETER_NODE_NAME ),
-				names( methodName, TestUtil.CROSS_PARAMETER_NODE_NAME )
-		);
-		assertCorrectPathNodeKinds(
-				violations,
-				kinds( ElementKind.METHOD, ElementKind.CROSS_PARAMETER ),
-				kinds( ElementKind.METHOD, ElementKind.CROSS_PARAMETER )
+		assertThat( violations ).containsOnlyPaths(
```
Comment (gunnarmorling) [inline tests/src/main/java/org/hibernate/beanvalidation/tck/tests/methodvalidation/ValidateParametersTest.java]: Shouldn't this be two paths here? I.e. how comes it's not failing?

File: tests/src/main/java/org/hibernate/beanvalidation/tck/tests/methodvalidation/ValidateReturnValueTest.java [hunk 5]
```diff
@@ -229,15 +227,13 @@ public void testValidationWithSeveralGroups() throws Exception {
...
-				violations,
-				names( methodName, TestUtil.RETURN_VALUE_NODE_NAME ),
-				names( methodName, TestUtil.RETURN_VALUE_NODE_NAME )
-		);
-		assertCorrectPathNodeKinds(
-				violations,
-				kinds( ElementKind.METHOD, ElementKind.RETURN_VALUE ),
-				kinds( ElementKind.METHOD, ElementKind.RETURN_VALUE )
+		assertThat( violations ).containsOnlyPaths(
+				pathWith()
```
Comment (gunnarmorling) [inline tests/src/main/java/org/hibernate/beanvalidation/tck/tests/methodvalidation/ValidateReturnValueTest.java]: I have to admit the previous approach was a bit silly as it ignored ordering. Now it makes sure that the type and name etc. of the same path are correct. +1!

File: tests/src/main/java/org/hibernate/beanvalidation/tck/util/ConstraintViolationAssert.java [hunk 6]
```diff
@@ -45,10 +51,224 @@
...
+		assertCorrectConstraintTypes( actualConstraintTypes, expectedConstraintTypes );
+	}
+
+	/**
+	 * Asserts that the given list of constraint violation paths matches the list of expected property paths.
+	 *
+	 * @param violations The violation list to verify.
+	 * @param expectedPropertyPaths The expected property paths.
+	 */
+	public static void assertCorrectPropertyPaths(Set<? extends ConstraintViolation<?>> violations,
```
Comment (gunnarmorling) [inline tests/src/main/java/org/hibernate/beanvalidation/tck/util/ConstraintViolationAssert.java]: I'm a bit confused about this one. Isn't it what you wanted to get rid of?

File: tests/src/main/java/org/hibernate/beanvalidation/tck/tests/validation/CustomPropertyPathTest.java [hunk 7]
```diff
@@ -106,12 +106,12 @@ public void testAddingNodesInClassLevelConstraintKeepsInIterableKeyAndIndex() {
...
 				pathWith().property( "fooArray" ).property( "myNode8", true, null, 1, Object[].class, null ).property( "myNode9", true, "Foo", null ),
 				pathWith().property( "fooArray" ).property( "myNode10", true, null, 1, Object[].class, null ).property( "myNode11", true, null, null ).property( "myNode12" ),
-				pathWith().property( "fooSet" ).property( "myNode1", true, null, null, Iterable.class, 0 ),
-				pathWith().property( "fooSet" ).property( "myNode2", true, null, null, Iterable.class, 0 ).property( "myNode3" ),
-				pathWith().property( "fooSet" ).property( "myNode4", true, null, null, Iterable.class, 0 ).property( "myNode5", true, null, null ),
-				pathWith().property( "fooSet" ).property( "myNode6", true, null, null, Iterable.class, 0 ).property( "myNode7", true, null, 42 ),
-				pathWith().property( "fooSet" ).property( "myNode8", true, null, null, Iterable.class, 0 ).property( "myNode9", true, "Foo", null ),
-				pathWith().property( "fooSet" ).property( "myNode10", true, null, null, Iterable.class, 0 ).property( "myNode11", true, null, null ).property( "myNode12" ),
+				pathWith().property( "fooSet" ).property( "myNode1", true, null, null, Set.class, 0 ),
```
Comment (gunnarmorling) [inline tests/src/main/java/org/hibernate/beanvalidation/tck/tests/validation/CustomPropertyPathTest.java]: How comes the type has changed here?

Comment (gsmet) [inline tests/src/main/java/org/hibernate/beanvalidation/tck/tests/validation/CustomPropertyPathTest.java]: It used to be wrong and I fixed it (and fixed it in HV too).

File: tests/src/main/java/org/hibernate/beanvalidation/tck/util/ConstraintViolationAssert.java (hunk 1 above)
Comment (gsmet) [inline tests/src/main/java/org/hibernate/beanvalidation/tck/util/ConstraintViolationAssert.java]: No. It tests that the actualViolations parameter passed in the assertThat is not null. (it's a straight copy of what is in HV)

File: tests/src/main/java/org/hibernate/beanvalidation/tck/tests/constraints/builtinconstraints/FuturePastConstraintsTest.java (hunk 2 above)
Comment (gsmet) [inline tests/src/main/java/org/hibernate/beanvalidation/tck/tests/constraints/builtinconstraints/FuturePastConstraintsTest.java]: Not really, it started failing as the tests were coming from another package. So I decided to fix it by bringing them back here instead of relying on unrelated entities.

File: tests/src/main/java/org/hibernate/beanvalidation/tck/tests/constraints/builtinconstraints/DigitsConstraintTest.java (hunk 3 above)
Comment (gsmet) [inline tests/src/main/java/org/hibernate/beanvalidation/tck/tests/constraints/builtinconstraints/DigitsConstraintTest.java]: No, AFAIR, it has been removed in a later commit.

File: tests/src/main/java/org/hibernate/beanvalidation/tck/tests/methodvalidation/ValidateParametersTest.java (hunk 4 above)
Comment (gsmet) [inline tests/src/main/java/org/hibernate/beanvalidation/tck/tests/methodvalidation/ValidateParametersTest.java]: This is why we need to check the number of violations. containsOnly deduplicates the elements.

File: tests/src/main/java/org/hibernate/beanvalidation/tck/util/ConstraintViolationAssert.java (hunk 6 above)
Comment (gsmet) [inline tests/src/main/java/org/hibernate/beanvalidation/tck/util/ConstraintViolationAssert.java]: It got removed in a separate commit.

**Comment by gsmet** [comment]
@gunnarmorling I answered your comments. I don't see anything to fix on my side.

File: tests/src/main/java/org/hibernate/beanvalidation/tck/tests/methodvalidation/ValidateParametersTest.java (hunk 4 above)
Comment (gsmet) [inline tests/src/main/java/org/hibernate/beanvalidation/tck/tests/methodvalidation/ValidateParametersTest.java]: I suspect we should use http://joel-costigliola.github.io/assertj/core/api/org/assertj/core/api/AbstractIterableAssert.html#containsExactlyInAnyOrder(ELEMENT...) instead of containsOnly. I'm pretty sure it's a new thing in assertj and it wasn't existing when it was introduced.

I think it would be a nice change but for another PR :). Opening an issue for that.

Comment (gunnarmorling) [inline tests/src/main/java/org/hibernate/beanvalidation/tck/tests/methodvalidation/ValidateParametersTest.java]: Ok, filed [BVTCK-2^7](https://hibernate.atlassian.net/browse/BVTCK-128). Let's do it for the next release, would be a good improvement.
